logger.js

```diff
@@ -10,135 +10,142 @@
-		warn: 4,
-		error: 5,
-		fatal: 6,
-	};
-
-	config.level_abbr = config.level_abbr || {
-		trace: 'trc',
-		debug: 'dbg',
-		log: 'log',
-		info: 'inf',
-		warn: 'WRN',
-		error: 'ERR',
-		fatal: 'FTL',
-	};
-
-	config.filename = `${process.cwd()}/${config.filename || 'logs.log'}`;
-
-	config.errorLevel = config.errorLevel || 'log';
-
-	child_process.execSync(`mkdir -p ${path.dirname(config.filename)}`);
-	var log_file = fs.createWriteStream(config.filename, { flags: 'a' });
+		debug: 4,
+		query: 5,
+		trace: 10,
+		none: 99,
+	},
+	colors: {
+		fatal: 'yellow',
+		error: 'red',
+		warn: 'meganta',
+		info: 'blue',
+		debug: 'green',
+		query: 'green',
+		trace: 'cyan',
+		none: 'black',
+	},
+};
 
-	exports.setLevel = function(errorLevel) {
-		config.errorLevel = errorLevel;
-	};
+const logRootKeys = [
+	'level',
+	'timestamp',
+	'module',
+	'message',
+	'action',
+	'hostname',
+	'durationMs',
+];
+
+winston.addColors(customLevels.colors);
+
+const defaultLoggerConfig = {
+	filename: 'logs/lisk.log',
+	level: 'none',
+	consoleLevel: 'debug',
+};
 
-	function snipFragileData(data) {
-		for (var key in data) {
-			if (key.search(/passphrase|password/i) > -1) {
-				data[key] = 'XXXXXXXXXX';
+const formatInfo = format(info =>
+	Object.keys(info).reduce((accumulated, key) => {
+		if (logRootKeys.includes(key)) {
+			accumulated[key] = info[key];
+		} else {
+			if (!accumulated.meta) {
+				accumulated.meta = {};
 			}
+			accumulated.meta[key] = info[key];
 		}
-		return data;
-	}
-
-	Object.keys(config.levels).forEach(name => {
-		function log(message, data) {
-			var log = {
-				level: name,
-				timestamp: strftime('%F %T', new Date()),
-			};
+		return accumulated;
+	}, {}));
+
+const consoleFormat = info => {
+	let log = `[${info.level}] ${info.timestamp} <${info.module}>: ${
+		info.message
+	}`;
+	log += info.durationMs ? ` duration=${info.durationMs}ms` : '';
+	log += info.meta ? `\n\t ${JSON.stringify(info.meta)}` : '';
+	return log;
+};
 
-			if (message instanceof Error) {
-				log.message = message.stack;
-			} else {
-				log.message = message;
-			}
+const injectKey = (key, value) =>
+	format(info => {
+		info[key] = value;
+		return info;
+	});
 
-			if (data && util.isObject(data)) {
-				log.data = JSON.stringify(snipFragileData(data));
-			} else {
-				log.data = data;
-			}
+const consoleLog = (level, module) =>
+	new winston.transports.Console({
+		level,
+		levels: customLevels.levels,
+		format: format.combine(
+			format.colorize(),
+			injectKey('module', module)(),
+			format.timestamp(),
+			formatInfo(),
+			format.printf(consoleFormat)
+		),
+	});
 
-			log.symbol = config.level_abbr[log.level]
-				? config.level_abbr[log.level]
-				: '???';
+const fileLog = (level, filename, module) =>
+	new winston.transports.File({
+		level,
+		filename,
+		format: format.combine(
+			format.timestamp(),
+			injectKey('module', module)(),
+			injectKey('hostname', os.hostname())(),
+			formatInfo(),
+			format.json()
+		),
```

Comment:
I am not sure about this structure, does that means every log entry will make 6 additional function calls? Or minimum 3 additional calls to `injectKeys` and `formatInfo`. I feel that's a performance degrade for the each log entry.

Reply:
It will go through all of the functions like middleware, but I think it will happen in any logger where we want to format the log before outputting it

Reply:
If that's happening in all logging tools then there is a right reason to create our own logger. This is too much of the computation power to be used unnecessary for logging.

Reply:
Even if we do, we will need similar operation, unless we have better way to inject globally into the logger.